src/encoding/encoding.rs

```diff
@@ -248,15 +298,19 @@ impl Encoding {
         self.draw_data
             .extend_from_slice(bytemuck::bytes_of(&DrawBeginClip::new(blend_mode, alpha)));
         self.n_clips += 1;
+        self.n_open_clips += 1;
     }
 
     /// Encodes an end clip command.
     pub fn encode_end_clip(&mut self) {
-        self.draw_tags.push(DrawTag::END_CLIP);
-        // This is a dummy path, and will go away with the new clip impl.
-        self.path_tags.push(PathTag::PATH);
-        self.n_paths += 1;
-        self.n_clips += 1;
+        if self.n_open_clips > 0 {
+            self.draw_tags.push(DrawTag::END_CLIP);
+            // This is a dummy path, and will go away with the new clip impl.
+            self.path_tags.push(PathTag::PATH);
+            self.n_paths += 1;
+            self.n_clips += 1;
+            self.n_open_clips -= 1;
+        }
```

Comment:
Should this else panic?

Reply:
In general, I don’t think libraries should ever intentionally panic though I would be open to returning an error here.

Reply:
I don't think this is a recoverable error, which is why I suggest a panic. Your use of the API is wrong if you ever reach this case. Maybe I'm misunderstanding

Reply:
It’s recoverable in the sense that we can avoid producing a corrupt encoding albeit likely with incorrect rendering.

I think it’s less misunderstanding and more difference of perspective. Since vello is likely to sit at a lower level of the stack, my preference is that malformed (potentially user generated) content should not kill the whole process without warning.

Reply:
I think the error should be reported somehow; if your user controlled data allows this case to happen, that's a bug in your translation of the user data.

Returning an error here would be awkward for consumers; it's the only place such an error can occur - maybe poisoning the `Scene/SceneBuilder` is good enough; panicing if there's such an error on dropping the builder, unless it was handled properly?

Reply:
Looking at other implementations: piet (and old piet-gpu) return an error in this case, but allow you to continue. Skia, notably, just [ignores](https://github.com/google/skia/blob/ad6b08ad714c22adbcee9f92f03e75949aaa925e/src/core/SkCanvas.cpp#L651) a restore operation if there are no saved layers on the stack.

Pragmatically, there are potential consumers of vello for which a bug in content translation causing a crash is not an acceptable result and we are subject to that constraint even if I conceptually agree with you on ensuring correctness.

Broader in scope, general robustness and error handling are things we want to tackle in a principled way. I think you mentioned this on the zulip and it's probably worth filing an issue to continue discussion.